### PR TITLE
Preserve microbatch sharding of carry inputs in Pipeline.

### PR DESCRIPTION
**Pull request by markblee**
Previously, carry inputs were explicitly padded to `M + N - 1` and scanned over directly. While simple, the dim cannot be sharded because it's being indexed by scan, and we also waste memory due to padding values.

### CODE REVIEW

File: axlearn/common/pipeline.py [hunk 1]
```diff
@@ -305,80 +310,99 @@ def vmap_fn(
...
+                # Note that vmap outputs are typically sharded over stages and may incur extra
+                # communication per-iteration (e.g. from broadcasting last stage outputs).
+                return carry_out, vmap_out
+
+            state_t0 = dict(
+                # Current loop iteration.
+                t=jnp.array(0, dtype=jnp.int32),
+                # [N, microbatch_size, ...].
+                carry_output_t_1=jax.tree_util.tree_map(
+                    lambda x: jnp.zeros((n,) + x.shape[1:]), carry
```
Comment (ptoulme-aws) [inline axlearn/common/pipeline.py:380]: see this PR: https://github.com/markblee/axlearn/pull/1

Comment (markblee) [reply axlearn/common/pipeline.py]: Thanks -- how can I reproduce this error? (Also added `x.dtype`, in case that resolves it.)

**Review by ruomingp — APPROVED** [review]

File: axlearn/common/pipeline.py (hunk 1 above)
Comment (ruomingp) [reply axlearn/common/pipeline.py]: Do we have a test for bfloat16?

Comment (markblee) [reply axlearn/common/pipeline.py]: Added a test that fails without the change.

**Review by ruomingp — APPROVED** [review]
Thanks for adding the test!
